### ISSUE TITLE
Chart is not defined etc.

### ISSUE THREAD

**Issue by gjb2048** [iandavidwild/moodle-block_completion_progress#3]
Hi Ian,

I'm reviewing chapter 7 of your book but can't get the code to compile as it should.  After sorting the LF issues I still get:

````
F:\moodledev\moodle33>grunt --root=/blocks/completion_progress/amd
>> Setting root to F:\moodledev\moodle33\blocks\completion_progress\amd

Running "startup" task

Running "eslint:amd" (eslint) task

F:\moodledev\moodle33\blocks\completion_progress\amd\src\chart_renderer.js
   71:13  error  'Chart' is not defined                                         no-undef
  132:31  error  'Chart' is not defined                                         no-undef
  150:33  error  String prototype is read only, properties should not be added  no-extend-native
  150:57  error  Trailing spaces not allowed                                    no-trailing-spaces
  155:37  error  'allData' is defined but never used                            no-unused-vars
  170:23  error  'chartData' is defined but never used                          no-unused-vars
  232:13  error  'Chart' is not defined                                         no-undef
  235:31  error  'Chart' is not defined                                         no-undef
  281:23  error  'chartData' is defined but never used                          no-unused-vars

✖ 9 problems (9 errors, 0 warnings)

Warning: Task "eslint:amd" failed. Use --force to continue.

Aborted due to warnings.

F:\moodledev\moodle33>
````

**Comment by gjb2048** [comment]
Additionally to solve the missing 'Chart' issue a long time ago I worked out in my Shoehorn theme that you could AMD define (require) the third party library if you put the minified version in the 'build' folder and its developer had added an AMD 'define' line like:

````
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.Chart = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
````
etc.

See: https://github.com/gjb2048/moodle-block_completion_progress/commit/dcdb7cfa073126b046c21257ded6e0bce7f760c4 - lots of changes due to 'tabs' not spaces, so the specific line is: https://github.com/gjb2048/moodle-block_completion_progress/commit/dcdb7cfa073126b046c21257ded6e0bce7f760c4#diff-3b7d73d3474253d142fda4d8064e905dR29 - the advantage of this is that you know that the third party lib will be only loaded when it is needed and it will not be after the custom JS that uses it because of now one require PHP line and AMD does the rest as it should.

But even with this the code will not compile:

````
F:\moodledev\moodle33>grunt --root=/blocks/completion_progress/amd
>> Setting root to F:\moodledev\moodle33\blocks\completion_progress\amd

Running "startup" task

Running "eslint:amd" (eslint) task

F:\moodledev\moodle33\blocks\completion_progress\amd\src\chart_renderer.js
  150:33  error  String prototype is read only, properties should not be added  no-extend-native
  155:37  error  'allData' is defined but never used                            no-unused-vars
  170:23  error  'chartData' is defined but never used                          no-unused-vars
  281:23  error  'chartData' is defined but never used                          no-unused-vars

✖ 4 problems (4 errors, 0 warnings)

Warning: Task "eslint:amd" failed. Use --force to continue.

Aborted due to warnings.

F:\moodledev\moodle33>
````

**Comment by iandavidwild** [comment]
Hi Gareth,

I have just built a minified version of the chart_renderer.js file and included it in the build so the plugin should work straight off (i.e. without having to turn JavaScript caching off).

There seemed to be a few issues with using chartjs and requirejs (e.g. [here](https://stackoverflow.com/questions/31288001/how-to-use-chart-js-with-angular-chart-using-requirejs)) so must admit rather than trying to get it to work (and experiencing similar issues to the ones you describe) I decided to load them separately (simply for the purposes of demonstrating the principles of using Chart.js in the book).
